core/primitives-core/src/profile.rs

```diff
@@ -19,154 +59,464 @@ impl Default for ProfileData {
 }
 
 impl ProfileData {
-    const EXT_START: usize = 1;
-    const ACTION_START: usize = ProfileData::EXT_START + ExtCosts::count();
-    const LEN: usize = 1 + ActionCosts::count() + ExtCosts::count();
-
     #[inline]
     pub fn new() -> Self {
-        let data = [0; ProfileData::LEN];
-        ProfileData { data }
+        let costs = DataArray([0; DataArray::LEN]);
+        ProfileData { data: costs }
     }
 
     #[inline]
     pub fn merge(&mut self, other: &ProfileData) {
-        for i in 0..ProfileData::LEN {
+        for i in 0..DataArray::LEN {
             self.data[i] = self.data[i].saturating_add(other.data[i]);
         }
     }
 
     #[inline]
     pub fn add_action_cost(&mut self, action: ActionCosts, value: u64) {
-        self.add_val(ProfileData::ACTION_START + action as usize, value);
+        self[Cost::ActionCost { action_cost_kind: action }] =
+            self[Cost::ActionCost { action_cost_kind: action }].saturating_add(value);
```

Comment:
Index is suggested by @matklad on this comment: https://github.com/near/nearcore/pull/4438#discussion_r674051345

So if having index it's natural to also have index_mut. 

I feel good for both `get()` and `index` flavor, but slightly prefer index, since it indicates constant time accessing.

Reply:
I see. Well, I can live with `index` implementation

Reply:
> I feel it is not the Rust way to abuse Index ([]) operator for something that is not a number or a range, 

> since it indicates constant time accessing.

I don't think either of this statements is true. Here's stdlib counter example for both:

```rust
fn main() {
    let map: std::collections::BTreeMap<_, _> = std::iter::once(("hello", 92)).collect();
	let value = map["hello"];
	println!("{}", value);
}
```

here, the `[]` is neither constant time, nor with range/integer. 

`[]` is a general operator for accessing a mapping in an *infallible* way, while `get` usually return options. In this case, we guarantee that the operation can never fail

That said, I don't have a strong opinion between Index or get, either is fine by me!

Reply:
@matklad Thanks! I didn't know BTreeMap can also be accessed with `[]`, but given so I think that was a bad decision. People get surprised of index is not a constant time operation, but breemap.get is fine and `get` is not surprising to be not O(1)